File: algorithms/cpp/onlineElection/onlineElection.cpp
// Source: https://leetcode.com/problems/online-election/
// Author: Miao Zhang
// Date:   2021-03-24

class TopVotedCandidate {
public:
    TopVotedCandidate(vector<int>& persons, vector<int>& times) {
        vector<int> votes(persons.size() + 1);
        int last_lead = persons.front();
        for (int i = 0; i < persons.size(); i++) {
            if (++votes[persons[i]] >= votes[last_lead]) {
                last_lead = persons[i];
            }
            leads_[times[i]] = last_lead;
        }
    }

    int q(int t) {
        return prev(leads_.upper_bound(t))->second;
    }
private:
    map<int, int> leads_; // time->lead
};

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate* obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj->q(t);
 */

File: algorithms/python/onlineElection/onlineElection.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/online-election/
# Author: Miao Zhang
# Date:   2021-03-24

class TopVotedCandidate:

    def __init__(self, persons: List[int], times: List[int]):
        self.leads = [[] for _ in range(len(times))]
        cnt = collections.Counter()
        last_lead = persons[0]
        for i, (t, p) in enumerate(zip(times, persons)):
            cnt[p] += 1
            if cnt[p] >= cnt[last_lead]:
                last_lead = p
            self.leads[i] = [t, last_lead]

    def q(self, t: int) -> int:
        return self.leads[bisect.bisect_right(self.leads, [t, float('inf')]) - 1][1]

